src/eos-updater-fetch.c

```diff
@@ -295,18 +299,451 @@ content_fetch_old (EosUpdater      *updater,
+                                                         path_relative_to_deployment,
+                                                         cancellable,
+                                                         &local_error);
+
+  if (!checkout_directory)
+    {
+      if (!g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        {
+          g_propagate_error (error, g_steal_pointer (&local_error));
+          return NULL;
+        }
+
+      return g_hash_table_new (NULL, NULL);
+    }
+
+  flatpak_ref_actions_table = eos_updater_util_flatpak_ref_actions_from_directory (path_relative_to_deployment,
+                                                                                   checkout_directory,
+                                                                                   FLATPAKS_IN_OSTREE_PRIORITY,
+                                                                                   cancellable,
+                                                                                   error);
+
+  /* Regardless of whether there was an error, we always want to remove
+   * the checkout directory at this point */
+  eos_updater_remove_recursive (checkout_directory, NULL);
+
+  if (!flatpak_ref_actions_table)
+    return NULL;
+
+  if (!ostree_repo_checkout_gc (repo, cancellable, error))
+    return NULL;
+
+  return g_steal_pointer (&flatpak_ref_actions_table);
+}
+
+/* As documented in OstreeRepoFinderResult, lower numbers indicate higher
+ * priority */
+static OstreeRepoFinderResult *
+highest_priority_repo_finder_result (OstreeRepoFinderResultv results)
+{
+  gint best_priority_score = G_MAXINT;
+  OstreeRepoFinderResult *best = NULL;
+  gsize i;
+
+  for (i = 0; results[i] != NULL; ++i)
+    {
+      if (results[i]->priority < best_priority_score)
+        {
+          best_priority_score = results[i]->priority;
+          best = results[i];
+        }
+    }
+
+  return best;
+}
+
+static gchar *
+find_first_flatpak_repo_for_collection_id (FlatpakInstallation        *installation,
+                                           const OstreeCollectionRef  *collection_ref,
+                                           GCancellable               *cancellable,
+                                           GError                    **error)
+{
+  g_autoptr(GFile) installation_directory = flatpak_installation_get_path (installation);
+  g_autoptr(GFile) repo_directory = g_file_get_child (installation_directory, "repo");
+  g_autoptr(GMainContext) context = g_main_context_new ();
+  g_autoptr(OstreeRepo) repo = ostree_repo_new (repo_directory);
+  g_autoptr(GAsyncResult) find_result = NULL;
+  g_auto(OstreeRepoFinderResultv) allocated_results = NULL;
+  OstreeCollectionRef const * collection_refs_to_fetch[2] = { collection_ref, NULL };
+
+  g_main_context_push_thread_default (context);
+
+  if (!ostree_repo_open (repo, cancellable, error))
+    return NULL;
+
+  ostree_repo_find_remotes_async (repo,
```

Comment:
Hmmm, I'm confused. I thought the reason why we wanted to specify a collection ID was that in flatpak land, `flatpak_remote_name == ostree_remote_name` and this was a bad thing. Are you saying that we can just use a flatpak remote name directly and that will be translated into a collection ID internally? What is the point of making all the configuration files take collection ID's in that case?

Reply:
In the meantime I'll reimplement using the above suggestion, thanks.

Reply:
`flatpak_remote_name == ostree_remote_name` always holds. We can use a flatpak remote name directly and that will be translated into a collection ID internally, indeed. There is no immediate need for the configuration files to use collection IDs: this could all work just as well if they used remote names. Over time, I would like things to move to using collection IDs (since they’re more globally unique than remote names), so the idea here was to use them from the start to avoid the need for a checkpoint in future to add them.

Using collection IDs here does mean that we can’t add anything to the install list until https://phabricator.endlessm.com/T19363 and https://phabricator.endlessm.com/T19487 have been fixed. These tasks add the `collection-id=` keys to the repo config files on new and already-installed images.

Since I want to be cautious about enabling P2P updates for everyone, then perhaps we _should_ use remote names in the autoinstall list (but *also* support collection IDs, so that we don’t need a checkpoint to add support for them in future). What do you think?